.github/copilot-instructions.md

```diff
@@ -0,0 +1,334 @@
+```bash
+# Test in specific shell
+bash -c "source nvm.sh && nvm --version"
+zsh -c "source nvm.sh && nvm --version"
+dash -c ". nvm.sh && nvm --version"
+```
+
+### Shell-Specific Considerations
+- **zsh**: Requires `setopt local_options shwordsplit` for word splitting
+- **dash**: Limited feature set, avoid bash-specific syntax
+- **ksh**: Some features may not work, primarily for compatibility testing
+
+## CI Environment Details
+
+### GitHub Actions Workflows
+
+#### `.github/workflows/tests.yml`
+- Runs test suite across multiple shells and test suites
+- Uses `script` command for proper TTY simulation
+- Matrix strategy covers shell × test suite combinations
+- Excludes install_script tests from non-bash shells
+
+#### `.github/workflows/shellcheck.yml`
+- Lints all shell scripts using shellcheck
+- Tests against multiple shell targets (bash, sh, dash, ksh)
+- Uses Homebrew to install latest shellcheck version
+
+#### `.github/workflows/lint.yml`
+- Runs additional linting and formatting checks
+- Validates documentation and code style
+
+### Travis CI (Legacy)
+- Configured in `.travis.yml`
+- Tests on multiple Ubuntu versions
+- Installs shell environments via apt packages
+
+### CI Test Execution
+```bash
+# Simulate CI environment locally
+export TRAVIS_BUILD_DIR=""  # Disable Travis-specific logic
+export GITHUB_ACTIONS=""    # Disable GitHub Actions logic
+make test
+```
+
+## Setting Up shellcheck Locally
+
+### Installation
+
+#### macOS (Homebrew)
+```bash
+brew install shellcheck
+```
+
+#### Ubuntu/Debian
+```bash
+sudo apt-get install shellcheck
+```
+
+#### From Source
+```bash
+# Download from https://github.com/koalaman/shellcheck/releases
+wget https://github.com/koalaman/shellcheck/releases/download/latest/shellcheck-latest.linux.x86_64.tar.xz
+tar -xf shellcheck-latest.linux.x86_64.tar.xz
+sudo cp shellcheck-latest/shellcheck /usr/local/bin/
+```
+
+### Usage
+
+#### Lint Main Files
+```bash
+shellcheck -s bash nvm.sh
+shellcheck -s bash install.sh  
+shellcheck -s bash nvm-exec
+shellcheck -s bash bash_completion
+```
+
+#### Lint Across Shell Types
+```bash
+shellcheck -s sh nvm.sh      # POSIX sh
+shellcheck -s bash nvm.sh    # Bash extensions
+shellcheck -s dash nvm.sh    # Dash compatibility
+shellcheck -s ksh nvm.sh     # Ksh compatibility
+```
+
+#### Common shellcheck Directives in nvm
+- `# shellcheck disable=SC2039` - Allow bash extensions in POSIX mode
+- `# shellcheck disable=SC2016` - Allow literal `$` in single quotes
+- `# shellcheck disable=SC2001` - Allow sed usage instead of parameter expansion
+- `# shellcheck disable=SC3043` - Allow `local` keyword (bash extension)
+
+### Fixing shellcheck Issues
+1. **Quoting**: Always quote variables: `"${VAR}"` instead of `$VAR`
+2. **POSIX compliance**: Avoid bash-specific features in portable sections
+3. **Array usage**: Use `set --` for positional parameters instead of arrays
+4. **Local variables**: Mark as bash-specific or avoid in POSIX functions
```

Comment:
this isn't accurate? local variables are declared with `local FOO` and then initialized on the next line (the latter is for ksh support)